DigitalLearningSolutions.Data/DataServices/UserDataService/UserCentreDetailsDataService.cs

```diff
@@ -0,0 +1,96 @@
+﻿namespace DigitalLearningSolutions.Data.DataServices.UserDataService
+{
+    using System.Collections.Generic;
+    using System.Data;
+    using Dapper;
+    using DigitalLearningSolutions.Data.Enums;
+    using DigitalLearningSolutions.Data.Exceptions;
+
+    public partial class UserDataService
+    {
+        public void CreateOrUpdateUserCentreDetails(
+            int userId,
+            int centreId,
+            string email,
+            IDbTransaction? transaction = null
+        )
+        {
+            if (EmailIsInUseByOtherUser(userId, email))
+            {
+                throw new DelegateCreationFailedException(DelegateCreationError.EmailAlreadyInUse);
+            }
+
+            var userCentreDetailsValues = new
+            {
+                userId,
+                centreId,
+                email,
+            };
+
+            var detailsAlreadyExist = connection.QuerySingle<bool>(
+                @"SELECT CASE
+                        WHEN EXISTS (SELECT ID FROM UserCentreDetails
+                            WHERE UserID = @userId AND CentreID = @centreId)
+                        THEN 1
+                        ELSE 0
+                        END",
+                new { userId, centreId }
+                );
+
+            if (detailsAlreadyExist)
+            {
+                connection.Execute(
+                    @"UPDATE UserCentreDetails
+                    SET Email = @email
+                    WHERE userID = @userId AND centreID = @centreId",
+                    userCentreDetailsValues,
+                    transaction);
+            }
+            else
+            {
+                connection.Execute(
+                    @"INSERT INTO UserCentreDetails
+                    (
+                        UserId,
+                        CentreId,
+                        Email
+                    )
+                    VALUES
+                    (
+                        @userId,
+                        @centreId,
+                        @email
+                    )",
+                    userCentreDetailsValues,
+                    transaction
+                );
+            }
+        }
+
+        public bool EmailIsInUseByOtherUser(int userId, string email)
+        {
+            return connection.QuerySingle<bool>(
+                @"SELECT CASE
+                        WHEN EXISTS (SELECT ID FROM USERS WITH (UPDLOCK, TABLOCKX, HOLDLOCK) WHERE PrimaryEmail = @email AND ID <> @userId)
+                        OR EXISTS (SELECT ID FROM UserCentreDetails d WITH (UPDLOCK, TABLOCKX, HOLDLOCK)

```

Comment:
I'm hoping that we don't need an exclusive lock here. Hoping Mike McLean can give me some answers... Watch this space.

Reply:
Talked to Mike. I don't think the case of two people updating their email to the same thing in between when the check happens and when the update happens is something we have to worry about. We have around 500,000 users, and suppose they each change their email twice a year, and 10% the time that email is the same as someone else's email (very generous), then we have an average of one conflict every ~300s. For the checks/updates to interleave they'd have to be within 100ms of each other, probably. Realistically, it's not going to happen.

So let's check that no other user is using the email just before updating, but not worry about locking.